agent/config/file_watcher.go

```diff
@@ -0,0 +1,222 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/hashicorp/go-hclog"
+)
+
+const timeoutDuration = 200 * time.Millisecond
+
+type FileWatcher struct {
+	watcher          *fsnotify.Watcher
+	configFiles      map[string]*watchedFile
+	EventsCh         chan *WatcherEvent
+	logger           hclog.Logger
+	reconcileTimeout time.Duration
+	cancel           context.CancelFunc
+	done             chan interface{}
+}
+
+type watchedFile struct {
+	modTime time.Time
+}
+
+type WatcherEvent struct {
+	Filename string
+}
+
+func NewFileWatcher(configFiles []string, logger hclog.Logger) (*FileWatcher, error) {
+	ws, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	cfgFiles := make(map[string]*watchedFile)
+
+	w := &FileWatcher{
+		watcher:          ws,
+		logger:           logger.Named("file-watcher"),
+		configFiles:      cfgFiles,
+		EventsCh:         make(chan *WatcherEvent),
```

Comment:
You probably want this to a the very least be a 1-buffered channel to give it just a bit of concurrency instead of forcing a hard sync between the event channel here and the thing on the other side.

Though really this is going to be gated on how often folks add/remove config files in the config directory, which is going to be on the scale of human-initiated quantities so you could make an argument for just picking a small number like 5-10 and calling that good for now.

Reply:
OK. Usually if there's some sort of handoff between two goroutines I tend to prefer a 1-buffered chan to an un-buffered channel as it breaks the hard interlock. If you've already thought about the hard interlock scenario then it's fine.